aten/src/ATen/core/type.cpp

```diff
@@ -445,6 +445,20 @@ CAFFE2_API TypePtr tryEvalTypeVariables(TypePtr type, std::unordered_map<std::st
   }
 }
 
+CAFFE2_API bool elementTypeCanBeInferredFromMembers(const TypePtr& elem_type) {
```

Comment:
`elementTypeCanBeInferredFromMembers` makes sense in the context of type contains an element type. This function no longer has any reference to a container. 

maybe something like: `typeCanBeHetererogenous`, or `typeCanBePolymorphic`. I'm not sure what the best term is here (i also don't care that much). @bhosmer any thoughts on the best phrasing here ?

Reply:
How about `typeIsFinal` - "final" in the same sense as "final class" (no subtypes). It's short but also has type-theoretic fanciness :) 

To be totally clear about the function we're naming: saying
"`typeIsFinal(T)` returns true for some type `T`" 
is the same as saying, 
"given `mylist: List<T>` for some concrete type `T`, `typeIsFinal(T)` will return true iff elements of `mylist` are guaranteed to be exactly of type `T` (not a subtype of `T`)."